PHP 8 support

edudobay opened the pull request:
This PR allows Slim to be installed in PHP 8 and passes all the tests and checks.

History: I've raised some points in https://github.com/slimphp/Slim/issues/2977#issuecomment-703194522 and since that time I've worked around other repositories to make this possible. This is also my first open-source contribution to a major project :)

**Test fixes:**

* Added a test for an uncovered branch of `MiddlewareDispatcher->addDeferred`. PHPUnit 8 was miscalculating coverage – see https://coveralls.io/builds/33553304/source?filename=Slim/MiddlewareDispatcher.php.

**My questions:**

* I implemented `disableXmlEntityLoader` as a `private static method`. Is this okay? I thought about making it `protected` or extracting it to a class, but I'm unsure about creating new "contracts".
* Coverage – conditional call to `libxml_disable_entity_loader`. Is it okay to just `@codeCoverageIgnore` some part that will only run in PHP 7 or PHP 8? How should situations like this be handled?

**Dependencies:**  I have gone through all dependencies that are blocking the PHP 8 upgrade and temporarily linked to development branches where the PHP 8 issues are solved. Surely this must be addressed before this can be merged (this is isolated in a single commit that can be removed). Those are the open PRs, some by me and some by others:

* [x] https://github.com/slimphp/Slim-Psr7/pull/169
* [x] https://github.com/slimphp/Slim-Http/pull/142
* [x] https://github.com/php-fig/http-message-util/pull/19
* [x] https://github.com/Nyholm/psr7-server/pull/40
* [x] https://github.com/laminas/laminas-diactoros/pull/46

l0gicgate commented:
@edudobay thank you for this PR

> I implemented disableXmlEntityLoader as a private static method. Is this okay? I thought about making it protected or extracting it to a class, but I'm unsure about creating new "contracts".

I think that we could just make that method `protected` but does not need to be `static` unless I'm missing something

> Coverage – conditional call to libxml_disable_entity_loader. Is it okay to just @codeCoverageIgnore some part that will only run in PHP 7 or PHP 8? How should situations like this be handled?

I don't think there's any way around this one in this particular case. We would have to overload the `LIBXML_VERSION` constant somehow to be able to stub the behavior.

> Dependencies: I have gone through all dependencies that are blocking the PHP 8 upgrade and temporarily linked to development branches where the PHP 8 issues are solved. Surely this must be addressed before this can be merged (this is isolated in a single commit that can be removed). Those are the open PRs, some by me and some by others:

This particular issue bothers me a bit because we're going to have to do this work all over again when all the other libraries have official releases to support PHP 8. I would rather just do all the work at once than do it twice. We can just leave this PR open for now until the other libraries release official support for PHP 8.

edudobay commented:
@l0gicgate Sure, I'll make it `protected` then. But an instance method would not work inside the `static` anonymous function. I suppose that function is static for performance reasons. When I was implementing these changes, I benchmarked several kinds of function calls and I got differences up to 10x in memory usage — even though I don't think this is performance-critical code, I would be careful not to introduce a 10x factor in resource usage. I'll redo these benchmarks (I think I threw it away at the time), but I'll be happy to discuss other solutions too.

Regarding dependencies, I'll just watch them for releases and ping a comment when they're all ready.

edudobay commented:
Hi! I've updated the source branch and fixed some PHPStan issues that surfaced recently. I'm not sure if my solution was the way to go — the type hints that could go wrong are the ones for `Closure::bindTo()`, which can fail if called on a closure with static scope.

edudobay commented:
Hi @l0gicgate. All set for external projects! When https://github.com/slimphp/Slim-Psr7/pull/169 is merged and release tags are ready, we can remove the "TEMP" commit to go back to tagged release constraints.

I guess we will need to wait a few days for an official PHP 8.0 tag to be available in Travis CI, so we can fix that too, right? I'm not sure if we should otherwise remove "allowed to fail" from the nightly build.

l0gicgate commented:
> I guess we will need to wait a few days for an official PHP 8.0 tag to be available in Travis CI, so we can fix that too, right? I'm not sure if we should otherwise remove "allowed to fail" from the nightly build.

Once they release the official PHP 8.0 tag we can add it to the CI builds and leave the nightly build as "allowed to fail" as we will need it to examine the behavior of future versions on CI e.g: PHP 9 (long time from now)

edudobay commented:
@l0gicgate Great, let me know if I can help with anything!

l0gicgate commented:
@edudobay see the releases below:

[Slim-Psr7 1.3.0 Release](https://github.com/slimphp/Slim-Psr7/releases/tag/1.3.0)
[Slim-Http 1.2.0 Release](https://github.com/slimphp/Slim-Http/releases/tag/1.2.0)

We should be good to go for this PR after!

edudobay commented:
@l0gicgate I've updated composer.json to require these latest `slim/http` and `slim/psr7` if that's ok.

l0gicgate commented on Slim/CallableResolver.php:
```diff
@@ -183,6 +183,7 @@ private function bindToContainer(callable $callable): callable
         }
         if ($this->container && $callable instanceof Closure) {
             $callable = $callable->bindTo($this->container);
+            /** @var Closure $callable */
```
can we put the comment above the variable?

edudobay replied:
sure! Sorry I missed these little things :)

l0gicgate commented on Slim/Middleware/BodyParsingMiddleware.php:
```diff
@@ -188,4 +188,17 @@ protected function getMediaType(ServerRequestInterface $request): ?string
+
+    protected static function disableXmlEntityLoader(bool $disable): bool
+    {
+        if (\LIBXML_VERSION >= 20900) {
```
Let's add this to the imports. We don't do `\` in the codebase

```php
use LIBXML_VERSION;
```

l0gicgate commented on Slim/MiddlewareDispatcher.php:
```diff
@@ -288,6 +288,7 @@ public function addCallable(callable $middleware): self
 
         if ($this->container && $middleware instanceof Closure) {
             $middleware = $middleware->bindTo($this->container);
+            /** @var Closure $middleware */
```
Let's put the comment above the variable

l0gicgate reviewed (commented):
Just a few little things then I'll merge

edudobay commented:
@l0gicgate I've fixed the issues you mentioned!

l0gicgate reviewed (approved):
Thank you for this @edudobay!
